presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java

```diff
@@ -379,27 +382,32 @@ public CounterStat getHighMemorySplitSource()
                     partitions);
 
             ImmutableList.Builder<HivePartitionMetadata> results = ImmutableList.builder();
+            Map<String, Set<String>> partitionsNotReadable = new HashMap<>();
+            int unreadablePartitionsSkipped = 0;
             for (HivePartition hivePartition : partitionBatch) {
                 Partition partition = partitions.get(hivePartition.getPartitionId());
                 if (partition == null) {
                     throw new PrestoException(GENERIC_INTERNAL_ERROR, "Partition not loaded: " + hivePartition);
                 }
-                String partName = makePartName(table.getPartitionColumns(), partition.getValues());
+                String partitionName = makePartName(table.getPartitionColumns(), partition.getValues());
                 Optional<EncryptionInformation> encryptionInformation = encryptionInformationForPartitions.map(metadata -> metadata.get(hivePartition.getPartitionId()));
 
                 if (!isOfflineDataDebugModeEnabled(session)) {
                     // verify partition is online
-                    verifyOnline(tableName, Optional.of(partName), getProtectMode(partition), partition.getParameters());
+                    verifyOnline(tableName, Optional.of(partitionName), getProtectMode(partition), partition.getParameters());
 
                     // verify partition is not marked as non-readable
-                    String partitionNotReadable = partition.getParameters().get(OBJECT_NOT_READABLE);
-                    if (!isNullOrEmpty(partitionNotReadable)) {
+                    String reason = partition.getParameters().get(OBJECT_NOT_READABLE);
+                    if (!isNullOrEmpty(reason)) {
                         if (!shouldIgnoreUnreadablePartition(session) || !partition.isEligibleToIgnore()) {
-                            throw new HiveNotReadableException(tableName, Optional.of(partName), partitionNotReadable);
+                            throw new HiveNotReadableException(tableName, Optional.of(partitionName), reason);
+                        }
+                        if (partitionsNotReadable.size() < 4) {
+                            partitionsNotReadable.putIfAbsent(reason, new HashSet<>(ImmutableSet.of(partitionName)));
+                            if (partitionsNotReadable.get(reason).size() < 4) {
```

Comment:
To be consistent with the following `> 2`, these two should be `<= 2` right?

Reply:
@highker , what I meant to type was <=3 in this section, and >2 in the other section. The reason the numbers are inconsistent is because by adding at most N partitions to partitionsNotReadable, then checking in the other section if there is more than N-1 partitions, I can add N-1 partitions to the error message and decide whether a trailing "..." is needed or not.